### ISSUE TITLE
Project status

### ISSUE THREAD

**Issue by KaneRoot** [parke/lxroot#2]
Hello. I saw your repository and I like the project and consider using your program for a project of mine (involving compiling stuff on an alpine environment).

I wonder what do you have in mind for the future?
Do you currently use it, or know any project that does?

Currently there is hardly any documentation. A man page could be great, or at least an up-to-date readme.

Makefile could use a little help. I like to add a little `-Wextra` to catch a few more things that could go wrong.

I skimmed the project, and I saw there is a lot of code. A big chunk of it is dead-code, commented.
I wonder what parts of your code are relevant to the core functionality, C++ really seems overkill.
Could a few lines of shell, or a makefile, do the trick?
1000+ lines of C++ calling ls, cp, mkdir, cd, bind, mount, umount… do seems like shell scripting anyway. Did I miss something?

Thanks for reading my comment/suggestion/encouragement/help/rant.

**Comment by parke** [comment]
> Hello. I saw your repository and I like the project and consider using your program for a project of mine (involving compiling stuff on an alpine environment).

Nifty!  May I ask, how did you learn of lxroot?

> I wonder what do you have in mind for the future?

Can you be more specific?  My goal is to expose the available user namespace capabilities in a command-line tool that is easy for the user to reason about.

I have done much work on lxroot since the version that is published to github.  I did not publish those changes as publishing takes time, and no one was using lxroot.  So your contacting me will inspire me to publish a new version.

> Do you currently use it, or know any project that does?

I do use lxroot constantly.  I run a variety of desktop apps inside lxroot, including the Chromium web browser.  I run an Arch Linux userland on top of an Ubuntu host system running an Ubuntu kernel.

> Currently there is hardly any documentation. A man page could be great, or at least an up-to-date readme.

Indeed.  You, as a potential user, may inspire me to improve the documentation.

> Makefile could use a little help. I like to add a little `-Wextra` to catch a few more things that could go wrong.

Interesting.  `-Wextra` does more than `-Wall`?  (Maybe I already knew that.)  In general, I am in favor of activating as many compile time error messages as possible.
 
> I skimmed the project, and I saw there is a lot of code. A big chunk of it is dead-code, commented.

There is still some dead code in the unpublished version.  Dead code is typically code that is in the process of being permanently removed.  I've been refactoring things, so dead code may linger longer than is "professional".

Given that no one (other that me) was using lxroot, and I was still making sweeping changes to the command line interface (and possibly also to other aspects of lxroot), I was content to publish immature code.

> I wonder what parts of your code are relevant to the core functionality, C++ really seems overkill.

I started in C.  I moved to C++ to keep the code readable, while automating error handling (via templates), and improving error messages.

> Could a few lines of shell, or a makefile, do the trick?
> 1000+ lines of C++ calling ls, cp, mkdir, cd, bind, mount, umount… do seems like shell scripting anyway. Did I miss something?

The problem is that if you are not root, the **same process** (and even the **same program**?) that calls `unshare()` must also call `mount()`, `chroot()`, `pivot()`, etc.  Where this not this case, everything could be done via scripts.

> Thanks for reading my comment/suggestion/encouragement/help/rant.

Not a rant at all.  I'll try to push an update in the next 24-48 hours.  Thanks for contacting me!

**Comment by KaneRoot** [comment]
> Nifty! May I ask, how did you learn of lxroot?

Stalking a github account of a friend of mine, reading random stuff.
Not searching for that kind of tool in any way.


> > I wonder what do you have in mind for the future?
> 
> Can you be more specific? My goal is to expose the available user namespace capabilities in a command-line tool that is easy for the user to reason about.

Actually, just knowing that this is not a dead project is kinda enough.

> I have done much work on lxroot since the version that is published to github. I did not publish those changes as publishing takes time, and no one was using lxroot. So your contacting me will inspire me to publish a new version.

Nice, but don't rush things for me.

> > Do you currently use it, or know any project that does?
> 
> I do use lxroot constantly. I run a variety of desktop apps inside lxroot, including the Chromium web browser. I run an Arch Linux userland on top of an Ubuntu host system running an Ubuntu kernel.

Great. This reminds me of firejail, any thoughts on it? https://firejail.wordpress.com/


> > Currently there is hardly any documentation. A man page could be great, or at least an up-to-date readme.
> 
> Indeed. You, as a potential user, may inspire me to improve the documentation.

What I will do with your tool is to build packages. So I guess it should be a very simple usage to cover.


> > I wonder what parts of your code are relevant to the core functionality, C++ really seems overkill.
> 
> I started in C. I moved to C++ to keep the code readable, while automating error handling (via templates), and improving error messages.
> 
> > Could a few lines of shell, or a makefile, do the trick?
> > 1000+ lines of C++ calling ls, cp, mkdir, cd, bind, mount, umount… do seems like shell scripting anyway. Did I miss something?
> 
> The problem is that if you are not root, the **same process** (and even the **same program**?) that calls `unshare()` must also call `mount()`, `chroot()`, `pivot()`, etc. Where this not this case, everything could be done via scripts.

Let me try something : 
```bash
unshare mychroot
```
*mychroot*
```bash
mount
chroot
pivot
$(SHELL) # to let the user have a shell in the new environment
```

Couldn't something like this work? I would be happy to try to hack something like it in a few hours.
Actually, I didn't work on chroot environment for my builds up to this point, because it seemed too complicated to do correctly. But now, I'm motivated to find out what could be done with Linux namespaces.


> I'll try to push an update in the next 24-48 hours. Thanks for contacting me!

Don't rush, and you're welcome.

**Comment by parke** [comment]
> Actually, just knowing that this is not a dead project is kinda enough.

lxroot is not dead.  I did a lot of work, and made many changes during Sep through Dec of 2020.  I have more changes planned, but have not worked on lxroot since 1 Jan 2021.  I did do a lot of work after 1 Jan getting Debian packages to install inside an lxroot.  And a small amount of work install Arch packages.  `pacman` pretty much "just works" inside an lxroot, so not much work was required.  None of this package manager work required changes to lxroot.  It just required figuring out how to use lxroot to set up the user namespace so that the package managers would work.

I haven't been publishing the changes to lxroot because I have been making many changes, and some of those changes are only half-way-completed at present.  Despite this, lxroot works just fine.

> Great. This reminds me of firejail, any thoughts on it? https://firejail.wordpress.com/

I think about it like this:

chroot < lxroot < firejail < docker (or lxc or lxd)

However, only lxroot can function without root access.

I've never used firejail, but I have read parts of its documentation.

lxroot is focused solely on what can be done as non-root, without suid.  Firejail requires suid.  (At some point, I might add additional capabilities to lxroot when run as root, but I have no plans to add an suid mode.)

I thought about emailing the firejail mailing list and telling them about lxroot, but I never did, perhaps due to not having the time and energy.

Firejail does a lot more than lxroot.  Firejail's codebase is much larger and (imo) harder to read.

lxroot's primary goal is running (or compiling software).  In other words, security is not lxroot's primary goal.  Any security benefits of lxroot are simply a result of the fact that user namespaces provide security-ish capabilities, and lxroot provides access to those capabilities.

Another similar project is bubblewrap.

I want lxroot to "feel like" chroot to new users, with additional capabilities being optional and learnable one-at-a-time.

> What I will do with your tool is to build packages. So I guess it should be a very simple usage to cover.

I believe Alpine's build system builds packages as the "abuild" user, who is not root.  So a small amount of trickery might be necessary to get Alpine's build system running inside an lxroot.  (Or perhaps no trickery at all if you run lxroot as the abuild user on an Alpine system.)  At some point, I might want build Alpine packages inside an lxroot, but I have not needed to yet.

> Let me try something : [snip]

I'm not sure how to interpret what you are proposing.

If you run `unshare` as root, it may work.  Run as non-root, I think it won't work, but I could be wrong.

For your use case, you might be able to run unshare as root, do your mounting, chroot, and then `su` to a non-root user.

I wanted a solution that required no root access whatsoever, so I wrote lxroot.

> Actually, I didn't work on chroot environment for my builds up to this point, because it seemed too complicated to do correctly. 

chroot's are pretty simple.  The downside is that you cannot cleanly unmount them when you are done, as there are (typically) recursive bind mounts of /dev, /proc, and /sys, and the recursion prevents (clean) unmounting.  Running `unshare` as root might avoid this problem, as the recursive mounts would be inside the namespace, and the namespace will simply be garbage collected by the kernel when the last process exits.  (Or at least, that is what I would expect to happen.)

That said, Arch Linux installation process has an `arch-chroot` script that builds and enters the chroot environment, and it may be able to cleanly umount them upon exit.  I've never looked at it closely.

Since you're not in a hurry, I won't push an immature update.  But if you want to see a newer version of lxroot at some point, just let me know.  I expect to do more work on lxroot before the end of 2021.  Quite possibly even in the next few months.
